### PR TITLE
fix(init): guard against empty-vault deletion storm on partial volume wipe

### PR DESCRIPTION
**Pull request by aliasunder**
## Summary

- Adds a pre-sync guard in `init-first-sync` that detects the exact precondition for the 2026-08-15 deletion storm: vault volume empty + prior sync state present
- Uses a self-managed sentinel file (`.vault-synced`) on the `obsidian_config` volume — written after each successful sync, checked before the next. Sidesteps any dependency on obsidian-headless's undocumented internal file layout
- Guard fires regardless of `VAULT_NAME` — the deletion hazard depends on prior sync state (the device's file-tracking knowledge), not vault configuration
- Adds a callout warning in `deploy/remote/README.md` about never selectively deleting the vault volume

### How it works

1. After `ob sync` exits 0, `init-first-sync` writes `.vault-synced` to the config volume
2. On the next boot, before `ob sync` runs, the script checks: sentinel exists + vault has no visible files (dotfiles like `.obsidian/` excluded) → fatal exit
3. Wiping `obsidian_config` clears both the sentinel AND the device registration → fresh device → pure download (safe)

### Edge cases covered

| Scenario | Result |
|---|---|
| Fresh boot (all volumes new) | Guard passes → sync → sentinel written |
| Warm restart (all volumes kept) | Guard passes (vault has files) |
| **Vault wiped, config kept** | **Guard fires → FATAL** |
| Both volumes wiped | Guard passes → fresh device |
| Only `.obsidian/` in vault | Guard fires (no visible files) |
| First sync failed (no sentinel) | Guard passes |

## Test plan

- [x] 7 new test cases in `init-first-sync.test.ts` — all pass (21 total)
- [x] Mutation check: disabled the guard → exactly 3 tests fail (the guard tests)
- [x] `npm run build` — clean
- [x] `npm run lint` — 0 errors
- [x] `npm test` — 76 files, 2806 tests pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Added a startup safety check that prevents synchronization when a previously synced vault is unexpectedly empty, helping avoid accidental mass deletions across devices.
  * Synchronization now proceeds only after a successful initial sync; safely handled warning conditions preserve the current vault state.

* **Documentation**
  * Documented the new startup protection and recommended volume-management practices to prevent accidental data loss during deployments.

### CODE REVIEW

File: rootfs/etc/s6-overlay/scripts/init-first-sync [hunk 1]
```diff
@@ -23,6 +29,31 @@ if ! cd "$VAULT_PATH"; then
...
+# Pre-sync guard: an empty vault with a prior-sync sentinel means the vault
+# volume was wiped while the config volume (device state) was kept — syncing
+# would push mass deletions. The glob excludes dotfiles, so .obsidian/ alone
+# counts as empty.
+vault_has_visible_files() {
+  for entry in "$VAULT_PATH"/*; do
+    [ -e "$entry" ] && return 0
+  done
+  return 1
+}
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply rootfs/etc/s6-overlay/scripts/init-first-sync:45]: This is intentional — the guard detects freshly empty volumes (the 2026-08-15 incident scenario: volume recreated, all state gone), not partial file deletion. A visible non-hidden directory at the vault root IS evidence the vault is not a freshly mounted empty volume.

The "aborted rm -rf leaves an empty folder" scenario is not the deletion storm — if someone actively deleted files (even partially), sync propagating those deletions is correct behavior, not a storm. The storm is when the user never deleted anything — the volume itself was recreated, destroying all vault state, while the sync engine's file-tracking state (on the config volume) still says "I had 1,428 files."

Checking only for `-f` (regular files) would give false negatives on healthy vaults where the top level has only folders (common in Obsidian: `About Me/`, `Projects/`, `Daily Notes/` with no top-level `.md` files).

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: rootfs/etc/s6-overlay/scripts/init-first-sync [hunk 2]
```diff
@@ -23,6 +29,31 @@ if ! cd "$VAULT_PATH"; then
...
+
+if [ -f "$SYNC_SENTINEL" ] && ! vault_has_visible_files; then
+  echo "[obsidian-sync] ERROR: The vault is empty but this device has previously synced." >&2
+  echo "[obsidian-sync] Starting sync in this state would cause the sync engine to interpret" >&2
+  echo "[obsidian-sync] every previously-synced file as a local deletion, pushing mass" >&2
+  echo "[obsidian-sync] deletions to all connected devices." >&2
+  echo "[obsidian-sync] To restore the vault: re-attach the vault volume or restore from backup." >&2
+  echo "[obsidian-sync] To start fresh: remove the obsidian_config volume (re-registers the device)." >&2
+  exit 1
+fi
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply rootfs/etc/s6-overlay/scripts/init-first-sync:55]: This is intentional — "legitimately deleted all vault files on the server" is not a realistic scenario (migration means creating a new vault, not emptying the old one; consolidation means moving notes, not deleting them). The guard's error message already documents the escape hatch: "remove the obsidian_config volume (re-registers the device)."

Adding a `SYNC_ALLOW_EMPTY_VAULT` env var for this edge case adds a config axis, a code branch, a test matrix expansion, and documentation for a scenario that has never occurred and has no plausible trigger in normal Obsidian usage. The device re-registration cost of the config-volume wipe is appropriate friction for proceeding past a safety guard.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

**Comment by aliasunder** [comment]
@coderabbit review

File: deploy/remote/README.md [hunk 3]
```diff
@@ -368,6 +368,11 @@ docker compose down -v         # Compose
 docker stop vault-cortex
 ```
 
+> **Never delete just the vault volume while keeping `obsidian_config`.**
+> The sync engine would interpret the empty vault as mass local deletions
+> and push them to every connected device. Delete all volumes together
+> (`docker compose down -v`) or none.
+
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply deploy/remote/README.md:377]: Fixed — reworded to describe the guarded behavior (container refuses to start) rather than the unguarded scenario. Kept the severity context (mass deletions) so users understand why the guard exists.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*
